include/kernels/madwf_transfer.cuh

```diff
@@ -0,0 +1,152 @@
+            }
+          }
+        }
+      }
+      return out;
+    }
+
+    template <class storage_t, bool dagger_> struct Transfer5DArg : kernel_param<> {
+
+      using F = typename colorspinor_mapper<storage_t, 4, 3>::type;
+      using real = typename mapper<storage_t>::type;
+      using Vector = ColorSpinor<real, 3, 4>;
+      using matrix_t = typename transfer_5D_mapper<MadwfAcc::transfer_float, MadwfAcc::transfer_t>::type;
+
+      static constexpr bool dagger = dagger_;
+
+      F out;      // output vector field
+      const F in; // input vector field
+
+      const int Ls_out; // length of 5th dimension of the out field
+      const int Ls_in;  // length of 5th dimension of the in field
+
+      const int volume_4d_cb;
+
+      const matrix_t *wm_p; // points to the data that contains the transfer matrix
+
+      const int nParity;
+
+      Transfer5DArg(ColorSpinorField &out, const ColorSpinorField &in, const MadwfAcc::transfer_float *wm_p) :
+        kernel_param(dim3(out.VolumeCB() / out.X(4), out.X(4), out.SiteSubset())),
+        out(out),
+        in(in),
+        Ls_out(out.X(4)),
+        Ls_in(in.X(4)),
+        volume_4d_cb(in.VolumeCB() / in.X(4)),
+        wm_p(reinterpret_cast<const matrix_t *>(wm_p)),
+        nParity(in.SiteSubset())
+      {
+
+        if (volume_4d_cb != static_cast<int>(out.VolumeCB() / Ls_out)) {
+          errorQuda("Input and Output fields should have the same 4d volume: %d neq %d.\n", volume_4d_cb,
+                    static_cast<int>(out.VolumeCB() / Ls_out));
+        }
+
+        if (in.Nspin() != 4) errorQuda("nSpin = %d not supported", in.Nspin());
+        if (in.Ncolor() != 3) errorQuda("nColor = %d not supported", in.Ncolor());
+        if (out.Nspin() != 4) errorQuda("nSpin = %d not supported", out.Nspin());
+        if (out.Ncolor() != 3) errorQuda("nColor = %d not supported", out.Ncolor());
+
+        checkNative(in, out);
+      }
+    };
+
+    template <class Arg> struct Transfer5D {
+
+      const Arg &arg;
+      constexpr Transfer5D(const Arg &arg) : arg(arg) { }
+      static constexpr const char *filename() { return KERNEL_FILE; }
+
+      /**
+        @brief Apply the Ls_out-by-Ls_in matrix to the input vector
+        @param[in] parity Parity we are on
+        @param[in] x_b Checkerboarded 4-d space-time index
+        @param[in] s The output Ls dimension coordinate
+       */
+      __device__ __host__ inline void operator()(int x_cb, int s, int parity)
+      {
+        constexpr bool dagger = Arg::dagger;
+
+        using real = typename Arg::real;
+        using Vector = typename Arg::Vector;
+        using matrix_t = typename Arg::matrix_t;
+
+        const int Ls_in = arg.Ls_in;
+        const int Ls_out = arg.Ls_out;
+        const int volume_4d_cb = arg.volume_4d_cb;
+        const matrix_t *wm_p = arg.wm_p;
+
+        int thread_idx = target::thread_idx().y * target::block_dim().x + target::thread_idx().x;
+        SharedMemoryCache<real> cache;
```

Comment:
Just wondering if this can be written using `SharedMemoryCache<matrix_t>` instead.  Loading in each element, real by real, will mean the compiler can't use 64-bit or 128-bit instructions.  Could the matrix be assembled first in registers, and then copied to shared memory, and perhaps this would negate needing to accessing `data()` directly?

Reply:
Each thread loading one `matrix_t` would result in only `Ls_in * Ls_out` number of loads for a threadblock, which has too little parallelism for some block size. I have tried with `SharedMemoryCache<matrix_t>` and it results in a slow down when `Ls_in = 4`. Perhaps we can keep the code in this shape this for now?

Reply:
Ah, this makes sense.  Yep, fine to leave this as is then.